File: Leetcode 3301-3400/3388/3388.py
# Leetcode 3388: Count Beautiful Splits in an Array
# https://leetcode.com/problems/count-beautiful-splits-in-an-array/
# Solved on 1st of January, 2025
class Solution:
    def beautifulSplits(self, nums: list[int]) -> int:
        """
        Counts the number of beautiful splits in the given array.

        Args:
            nums: The input list of integers.
        Returns:
            The total count of beautiful splits.
        """
        n = len(nums)

        def getZ(start):
            length = n - start
            z = [0] * length
            l, r = 0, 0

            for i in range(1, length):
                if i <= r:
                    z[i] = min(r - i + 1, z[i - l])

                while i + z[i] < length and nums[start + z[i]] == nums[start + i + z[i]]:
                    z[i] += 1

                if i + z[i] - 1 > r:
                    l, r = i, i + z[i] - 1

            z[0] = length
            return z

        zMain = getZ(0)
        ans = 0

        for i in range(1, n - 1):
            isNums1Prefix = zMain[i] >= i
            zSuffix = getZ(i)

            for j in range(i + 1, n):
                len2 = j - i

                if isNums1Prefix and i <= len2:
                    ans += 1
                else:
                    if len2 <= n - j:
                        if zSuffix[j - i] >= len2:
                            ans += 1

        return ans

File: Leetcode 3301-3400/3388/3388.java
// Leetcode 3388: Count Beautiful Splits in an Array
// https://leetcode.com/problems/count-beautiful-splits-in-an-array/
// Solved on 1st of January, 2026
class Solution {
    /**
     * Counts the number of beautiful splits in the given array.
     *
     * @param nums The input array of integers.
     * @return The number of beautiful splits.
     */
    public int beautifulSplits(int[] nums) {
        int n = nums.length;
        int count = 0;

        int[] z = new int[n];
        for (int i = 1; i < n; i++) {
            int len = 0;
            while (i + len < n && nums[len] == nums[i + len]) {
                len++;
            }
            z[i] = len;
        }

        int[] nextLcp = new int[n + 1];
        int[] currLcp = new int[n + 1];

        for (int i = n - 2; i >= 1; i--) {
            for (int j = i + 1; j < n; j++) {
                if (nums[i] == nums[j]) {
                    currLcp[j] = 1 + nextLcp[j + 1];
                } else {
                    currLcp[j] = 0;
                }
            }

            for (int j = i + 1; j < n; j++) {
                boolean isBeautiful = false;

                if (i <= j - i && z[i] >= i) {
                    isBeautiful = true;
                }

                if (!isBeautiful && j - i <= n - j && currLcp[j] >= j - i) {
                    isBeautiful = true;
                }

                if (isBeautiful) {
                    count++;
                }
            }

            int[] temp = nextLcp;
            nextLcp = currLcp;
            currLcp = temp;
        }

        return count;
    }
}